File: library/src/main/java/org/incendo/interfaces/element/Element.java
package org.incendo.interfaces.element;

import org.bukkit.Material;
import org.bukkit.entity.Player;
import org.incendo.interfaces.click.ClickHandler;
import org.incendo.interfaces.drawable.Drawable;

public interface Element {

    Element EMPTY = new StaticElement(Drawable.drawable(Material.AIR));

    Drawable drawable();

    ClickHandler clickHandler();

    default CompletedElement complete(Player player) {
        return new CompletedElement(
            this.drawable().draw(player),
            this.clickHandler()
        );
    }

}

File: next/src/main/java/org/incendo/interfaces/next/element/Element.kt
package org.incendo.interfaces.next.element

import org.bukkit.Material
import org.incendo.interfaces.next.click.ClickHandler
import org.incendo.interfaces.next.drawable.Drawable

public interface Element {

    public companion object EMPTY : Element by StaticElement(Drawable.drawable(Material.AIR))

    public fun drawable(): Drawable

    public fun clickHandler(): ClickHandler
}
